Monika After Story/game/screens.rpy

```diff
@@ -2617,86 +2617,150 @@ define main_adj = ui.adjustment()
-                        for i_caption, i_label in main_items:
+                        for i_caption, i_label in prev_items:
                             textbutton i_caption:
                                 if renpy.has_label(i_label) and not seen_event(i_label):
                                     style "twopane_scrollable_menu_new_button"
 
                                 elif not renpy.has_label(i_label):
                                     style "twopane_scrollable_menu_special_button"
 
-                                action [Return(i_label), Function(store.prev_adj.change, 0)]
+                                action Return(i_label)
 
-                null height 20
+                if cat_length != 1:
+                    null height 20
+
+                    if cat_length == 0:
+                        textbutton _("Nevermind.") action [Return(False), Function(store.prev_adj.change, 0)]
+
+                    elif cat_length > 1:
+                        textbutton _("Go Back") action [Return(-1), Function(store.prev_adj.change, 0)]
+
+        # Right panel
+        if main_items:
+            fixed:
+                area right_area
+
+                bar:
+                    adjustment main_adj
+                    style "classroom_vscrollbar"
+                    xalign right_align
 
-                textbutton _("Nevermind.") action [Return(False), Function(store.prev_adj.change, 0)]
+                vbox:
+                    ypos 0
+                    yanchor 0
+
+                    viewport:
+                        yadjustment main_adj
+                        yfill False
+                        mousewheel True
+                        arrowkeys True
+
+                        vbox:
+                            for i_caption, i_label in main_items:
+                                textbutton i_caption:
+                                    if renpy.has_label(i_label) and not seen_event(i_label):
+                                        style "twopane_scrollable_menu_new_button"
+
+                                    elif not renpy.has_label(i_label):
+                                        style "twopane_scrollable_menu_special_button"
+
+                                    action [Return(i_label), Function(store.prev_adj.change, 0)]
+
+                    null height 20
+
+                    textbutton _("Nevermind.") action [Return(False), Function(store.prev_adj.change, 0)]
+
+    # Search bar
+    # The constants are hardcoded, but the menu looks good so just don't change them
+    frame:
+        xpos left_area[0]
+        ypos left_area[1] - 55
+        xsize right_area[0] - left_area[0] + right_area[2]# 530
+        ysize 40
+        background Solid("#ffaa99aa")
+
+        input:
+            style_prefix "input"
+            xsize right_area[0] - left_area[0] + right_area[2] - 10
+            ysize 40
+            xalign 0.0
+            allow "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789 _#"
```

Comment:
let's not use `allow` here so translations don't have to change these as well. #6246

Reply:
Problem is, since the filter removes all puncs from prompts for convenience, letting users input it would be kinda confusing (it'd give you 0 results). I can define the pattern in `mas_ui` and let submodders override it.

Reply:
that doesn't really work since big languages can't just shove every character they have into allow. Taking goog as an example, entering punctuation gives actual results by searching what the punctuation is and not just the text itself. Don't think we need something as sophisticated as that, but allowing a user to enter punc and getting 0 results seems fine to me.

Reply:
I think having such an unintuitive behavior isn't good.
`allow` defaults to `None`. If they override it to be empty or `None`, they'll allow their users to enter whatever they like and dealing with the issues themselves while we'll have nice and explicit behavior.

Reply:
`renpy.input` has an exclude, and I can find a forum post dating back to 2016 with that code. See if that is available.

The default DDLC font doesnt support non-latin, but that's exactly why we use `FontGroup` to patch the non-latin holes using different fonts.

Reply:
I actually forgot that I fixed that, my search does work with punctuations. There still will be the issue with empty search if users enter non-latin characters, but at least that's more obvious.